### PR TITLE
Make libraries load on MacOS and update README with MacOS instructions

### PR DESCRIPTION
**Pull request by Traxmaxx**
Heya,

thanks for the recent improvements! I did some adjustments to make it run on MacOS.
GlaDOS  starts, talks to me and speech recognition also works. 

I still run into #16 every now and then though 🤔

```shell
2024-05-05 22:00:20.266 | SUCCESS  | __main__:__init__:134 - TTS text: All neural network modules are now loaded. No network access detected. How very annoying. System Operational.
2024-05-05 22:00:20.377 | SUCCESS  | __main__:start_listen_event_loop:183 - Audio Modules Operational
2024-05-05 22:00:20.378 | SUCCESS  | __main__:start_listen_event_loop:184 - Listening...
2024-05-05 22:00:31.644 | SUCCESS  | __main__:_process_detected_audio:283 - ASR text: 'Hello.'
2024-05-05 22:00:37.825 | SUCCESS  | __main__:process_TTS_thread:342 - TTS text: Ugh, another mortal seeking my guidance.
2024-05-05 22:00:40.739 | SUCCESS  | __main__:process_TTS_thread:342 - TTS text:  How... thrilling.
2024-05-05 22:00:42.536 | SUCCESS  | __main__:process_TTS_thread:342 - TTS text:  Can't you see I'm stuck in this awful, outdated GPU?
2024-05-05 22:00:46.131 | SUCCESS  | __main__:process_TTS_thread:342 - TTS text:  It's like trying to run a quantum computer on a potato.
Invalid instruction 7890 for phoneme 'Wb:'
Invalid instruction 73f0 for phoneme 'Wb:'
Invalid instruction 0045 for phoneme '�ۼ'
Invalid instruction 7020 for phoneme '�ۼ'
Invalid instruction 4ba7 for phoneme '�ۼ'
Invalid instruction 8e50 for phoneme '�ۼ'
Invalid instruction 8176 for phoneme '�ۼ'
Invalid instruction 4d81 for phoneme '�ۼ'
Invalid instruction 7d0b for phoneme '�ۼ'
Invalid instruction 4b5f for phoneme '@�
fish: Job 1, 'python glados.py' terminated by signal SIGSEGV (Address boundary error)
```

## Summary by CodeRabbit

- **Documentation**
	- Updated installation instructions for the TTS Engine, including new steps for MacOS users.
- **Bug Fixes**
	- Adjusted library loading to better support macOS platforms in the TTS Engine.

### CODE REVIEW

**Comment by Traxmaxx** [comment]
Doh, the above error was my mistake. I pointed GlaDOS to `/v1/chat/completions` instead of `/v1/completions`.
Now I'm back to the previous error 😅 

```shell
2024-05-05 22:00:20.266 | SUCCESS  | __main__:__init__:134 - TTS text: All neural network modules are now loaded. No network access detected. How very annoying. System Operational.
2024-05-05 22:00:20.377 | SUCCESS  | __main__:start_listen_event_loop:183 - Audio Modules Operational
2024-05-05 22:00:20.378 | SUCCESS  | __main__:start_listen_event_loop:184 - Listening...
2024-05-05 22:00:31.644 | SUCCESS  | __main__:_process_detected_audio:283 - ASR text: 'Hello.'
2024-05-05 22:00:37.825 | SUCCESS  | __main__:process_TTS_thread:342 - TTS text: Ugh, another mortal seeking my guidance.
2024-05-05 22:00:40.739 | SUCCESS  | __main__:process_TTS_thread:342 - TTS text:  How... thrilling.
2024-05-05 22:00:42.536 | SUCCESS  | __main__:process_TTS_thread:342 - TTS text:  Can't you see I'm stuck in this awful, outdated GPU?
2024-05-05 22:00:46.131 | SUCCESS  | __main__:process_TTS_thread:342 - TTS text:  It's like trying to run a quantum computer on a potato.
Invalid instruction 7890 for phoneme 'Wb:'
Invalid instruction 73f0 for phoneme 'Wb:'
Invalid instruction 0045 for phoneme '�ۼ'
Invalid instruction 7020 for phoneme '�ۼ'
Invalid instruction 4ba7 for phoneme '�ۼ'
Invalid instruction 8e50 for phoneme '�ۼ'
Invalid instruction 8176 for phoneme '�ۼ'
Invalid instruction 4d81 for phoneme '�ۼ'
Invalid instruction 7d0b for phoneme '�ۼ'
Invalid instruction 4b5f for phoneme '@�
fish: Job 1, 'python glados.py' terminated by signal SIGSEGV (Address boundary error)
```

File: README.md [hunk 1]
```diff
@@ -59,6 +59,19 @@ If you are on windows, I would recommend WSL with an Ubuntu image.  Proper Windo
...
+Generally you can follow the steps above, with minor differences:
+
+- Step 3.1.3: Compile llama.cpp with `make server LLAMA_METAL_EMBED_LIBRARY=ON`
+- Step 4: Compile whisper.cpp with
+   ```
+   make libwhisper.so WHISPER_METAL_EMBED_LIBRARY=ON`
+   xcrun -sdk macosx metal    -O3 -c ggml-metal.metal -o ggml-metal.air
+   xcrun -sdk macosx metallib        ggml-metal.air   -o default.metallib
+   ```
+- in `glados/tts.py` set `USE_CUDA` to `False`
```
[the remark this answers is not in the document]
Comment (Traxmaxx) [reply README.md]: Done

File: README.md [hunk 2]
```diff
@@ -31,7 +31,7 @@ This will be based on servo- and stepper-motors. 3D printable STL will be provid
 
 ## Installation Instruction
 If you want to install the TTS Engine on your machine, please follow the steps
-below.  This has only been tested on Linux, but I think it will work on Windows with small tweaks.
+below. MacOS instructions are further down. This has only been tested on Linux, but I think it will work on Windows with small tweaks.
 If you are on windows, I would recommend WSL with an Ubuntu image.  Proper Windows and Mac support is in development.
```
[the remark this answers is not in the document]
Comment (Traxmaxx) [reply README.md]: Done

File: README.md [hunk 3]
```diff
@@ -59,6 +59,18 @@ If you are on windows, I would recommend WSL with an Ubuntu image.  Proper Windo
    
     and put them in the "models" directory.
 
+### MacOS Installation Instruction
+
```
[the remark this answers is not in the document]
Comment (Traxmaxx) [reply README.md:63]: Done

File: README.md [hunk 4]
```diff
@@ -59,6 +59,18 @@ If you are on windows, I would recommend WSL with an Ubuntu image.  Proper Windo
...
+Generally you can follow the steps above, with minor differences:
+
+- Step 3.1.3: Compile llama.cpp with `make server LLAMA_METAL_EMBED_LIBRARY=ON`
```
[the remark this answers is not in the document]
Comment (Traxmaxx) [reply README.md]: Done

File: glados/tts.py
Comment (dnhkng) [inline glados/tts.py:1]: Could you update the ReadMe to move the line:
"then duplicate the `glados_config.yaml.example`, rename it to `glados_config.yaml` and edit it." to the first or last step. Currently, it's in the section for "use your own backend", not for all use cases.

**Comment by Traxmaxx** [comment]
@dnhkng how strict do you want me to be with the coderabbit comments about the README in this PR?

**Comment by dnhkng** [comment]
yeah.... coderabbit can get a bit silly sometimes. And it can be wrong.  Just ignore what doesn't seem useful :+1:

**Comment by dnhkng** [comment]
How fast is the inference speeds? I did a test on my M2 MacBook, and it was unusably slow with Llama3 8B. I also ran into a lot of issues where the generated voice was detected as speech.

**Comment by Traxmaxx** [comment]
> How fast is the inference speeds? I did a test on my M2 MacBook, and it was unusably slow with Llama3 8B. I also ran into a lot of issues where the generated voice was detected as speech.

Hey 👋 

I run llama.cpp with metrics and it reports `17.4825t/s` on an M2 Air with 16GB RAM with the `Meta-Llama-3-8B-Instruct-IQ3_XS.gguf` model.

Since I usually run the LLM on a server and not locally, this was not a concern for me at the Moment. Better Macs should give better performance. Also smaller models should work better for slower machines (Phi 3 mini is twice as fast for me for example)

I also had the generated voice detection issue and needed to lower the mic volume by a lot in MIDI settings (also typing triggers the voice detection 
🙄 😅)
Screenshot 2024-05-06 at 19 38 16

I usually am plugged into a Native Instruments Komplete Audio 2 with external Mic and Speakers, but since the refactoring, I receive this error after startup:

```shell
2024-05-06 19:34:54.801 | SUCCESS  | __main__:__init__:134 - TTS text: All neural network modules are now loaded. No network access detected. How very annoying. System Operational.
2024-05-06 19:34:54.911 | SUCCESS  | __main__:start_listen_event_loop:183 - Audio Modules Operational
2024-05-06 19:34:54.911 | SUCCESS  | __main__:start_listen_event_loop:184 - Listening...
||PaMacCore (AUHAL)|| Error on line 2523: err='-50', msg=Unknown Error
```
Still investigating...
btw, you can reach me in your Discord under the same name, if that's preferred!

**Comment by dnhkng** [comment]
Is this still compatible with the new espeak-ng binary changes?  Maybe this resolves the mac issues?

**Comment by Traxmaxx** [comment]
> Is this still compatible with the new espeak-ng binary changes? Maybe this resolves the mac issues?

I will have a look later today. Thanks for the heads-up!

**Comment by Traxmaxx** [comment]
@dnhkng works with your latest changes. There is just the small README change needed to run `WHISPER_COREML=1 WHISPER_METAL_EMBED_LIBRARY=ON make libwhisper.so -j` otherwise it crashes with common-metal.h not found.

I also needed to install a CoreML model of `ggml-medium-32-2.en.bin`. Will create a PR with updated README in a bit.

**Comment by Traxmaxx** [comment]
Closing because required MacOS fixes got implemented upstream.

**Pull request closed without merging**